README.md
# osmcha-django

This is the Django backend API for [osmcha.org](https://osmcha.org/), a tool for reviewing and analyzing edits to [OpenStreetMap](https://openstreetmap.org/about).

The frontend code is in a separate repository, [osmcha-frontend](https://github.com/osmcha/osmcha-frontend). Bug reports and feature requests should be filed there by default, unless you're certain that your bug/request is specific to the backend.

License: BSD 2-Clause

OSMCha is an [OpenStreetMap US](https://openstreetmap.us/) project. You can support the project by [making a donation](https://openstreetmap.app.neoncrm.com/forms/osmcha).

## Settings

osmcha-django relies extensively on environment settings which **will
not work with Apache/mod_wsgi setups**. It has been deployed
successfully with both Gunicorn/Nginx and uWSGI/Nginx.

For configuration purposes, the following table maps the 'osmcha-django'
environment variables to their Django settings:

| Environment Variable                  | Django Setting                   | Development Default                        | Production Default                         |
| ------------------------------------- | -------------------------------- | ------------------------------------------ | ------------------------------------------ |
| DJANGO_CACHES                         | CACHES (default)                 | locmem                                     | redis                                      |
| DJANGO_DEBUG                          | DEBUG                            | True                                       | False                                      |
| DJANGO_SECRET_KEY                     | SECRET_KEY                       | CHANGEME!!!                                | raises error                               |
| DJANGO_SECURE_BROWSER_XSS_FILTER      | SECURE_BROWSER_XSS_FILTER        | n/a                                        | True                                       |
| DJANGO_SECURE_SSL_REDIRECT            | SECURE_SSL_REDIRECT              | n/a                                        | True                                       |
| DJANGO_SECURE_CONTENT_TYPE_NOSNIFF    | SECURE_CONTENT_TYPE_NOSNIFF      | n/a                                        | True                                       |
| DJANGO_SECURE_FRAME_DENY              | SECURE_FRAME_DENY                | n/a                                        | True                                       |
| DJANGO_SECURE_HSTS_INCLUDE_SUBDOMAINS | HSTS_INCLUDE_SUBDOMAINS          | n/a                                        | True                                       |
| DJANGO_SESSION_COOKIE_HTTPONLY        | SESSION_COOKIE_HTTPONLY          | n/a                                        | True                                       |
| DJANGO_SESSION_COOKIE_SECURE          | SESSION_COOKIE_SECURE            | n/a                                        | False                                      |
| DJANGO_DEFAULT_FROM_EMAIL             | DEFAULT_FROM_EMAIL               | n/a                                        | "osmcha-django \<noreply@example.com\>"    |
| DJANGO_SERVER_EMAIL                   | SERVER_EMAIL                     | n/a                                        | "osmcha-django \<noreply@example.com\>"    |
| DJANGO_EMAIL_SUBJECT_PREFIX           | EMAIL_SUBJECT_PREFIX             | n/a                                        | "\[osmcha-django\] "                       |
| DJANGO_CHANGESETS_FILTER              | CHANGESETS_FILTER                | None                                       | None                                       |
| PGUSER                                | DATABASES                        | None                                       | None                                       |
| PGPASSWORD                            | DATABASES                        | None                                       | None                                       |
| PGHOST                                | DATABASES                        | localhost                                  | localhost                                  |
| OAUTH2_OSM_KEY                        | SOCIAL_AUTH_OPENSTREETMAP_KEY    | None                                       | None                                       |
| OAUTH2_OSM_SECRET                     | SOCIAL_AUTH_OPENSTREETMAP_SECRET | None                                       | None                                       |
| DJANGO_ANON_USER_THROTTLE_RATE        | ANON_USER_THROTTLE_RATE          | None                                       | 30/min                                     |
| DJANGO_COMMON_USER_THROTTLE_RATE      | COMMON_USER_THROTTLE_RATE        | None                                       | 180/min                                    |
| DJANGO_NON_STAFF_USER_THROTTLE_RATE   | NON_STAFF_USER_THROTTLE_RATE     | 3/min                                      | 3/min                                      |
| OAUTH_REDIRECT_URI                    | OAUTH_REDIRECT_URI               | http://localhost:8000/oauth-landing.html   | http://localhost:8000/oauth-landing.html  |
| DJANGO_ENABLE_CHANGESET_COMMENTS      | ENABLE_POST_CHANGESET_COMMENTS   | False                                      | False                                      |
| DJANGO_OSM_COMMENTS_API_KEY           | OSM_COMMENTS_API_KEY             | ''                                         | ''                                         |

You can set each of these variables with:

    $ export VAR=VALUE

During the development, you can define the values inside your virtualenv
`bin/activate` file.

## Filtering Changesets

You can filter the changesets that will be imported by defining the
variable CHANGESETS_FILTER with the path to a GeoJSON file containing a
polygon with the geographical area you want to filter.

## Getting up and running

### Basics

The steps below will get you up and running with a local development
environment. We assume you have the following installed:

- pip
- virtualenv
- PostgreSQL

Before installing the python libraries, we need to install some packages
in the operational system:

    $ sudo ./install_os_dependencies.sh install

For the next step, make sure to create and activate a
[virtualenv](http://docs.python-guide.org/en/latest/dev/virtualenvs/),
then open a terminal at the project root and install the requirements
for local development:

    $ pip install -r requirements/local.txt

Create a local PostgreSQL database:

    $ createdb osmcha

Run `migrate` on your new database:

    $ python manage.py migrate

You can now run the `runserver_plus` command:

    $ python manage.py runserver_plus

Open up your browser to <http://127.0.0.1:8000/> to see the site running
locally.

### Setting Up Your Users

To create a **normal user account**, just go to Sign Up and fill out the
form. Once you submit it, you'll see a "Verify Your E-mail Address"
page. Go to your console to see a simulated email verification message.
Copy the link into your browser. Now the user's email should be verified
and ready to go.

To create a **superuser account**, use this command:

    $ python manage.py createsuperuser

For convenience, you can keep your normal user logged in on Chrome and
your superuser logged in on Firefox (or similar), so that you can see
how the site behaves for both kinds of users.

### How to login using the OAuth api

- Make a POST request to `<your_base_url>/api/v1/social-auth/` to
  receive the `oauth_token`, `oauth_token_secret` keys.
- Take the `oauth_token` and redirect the user to
  `https://www.openstreetmap.org/oauth/authorize?oauth_token=<oauth_token>`.
- You'll be redirected to the URL that you configured in your OSM OAuth
  key settings. That redirect url will contain the `oauth_verifier`
  param.
- Make another POST request to `<your_base_url>/api/v1/social-auth/` and
  send the `oauth_token`, `oauth_token_secret` and `oauth_verifier` as
  the data. You'll receive a token that you can use to make
  authenticated requests.
- The token key should be included in the Authorization HTTP header. The
  key should be prefixed by the string literal "Token", with whitespace
  separating the two strings. For example:
  `Authorization: Token 9944b09199c62bcf9418ad846dd0e4bbdfc6ee4b`.

### Frontend

[osmcha-frontend](https://github.com/osmcha/osmcha-frontend) is a web
interface that you can use to interact with the API. The web interface
can be deployed separately and configured to talk to this Django app
when making API requests. See the README in that repository for more
details.

### Feature creation endpoint

The feature creation endpoint allows only admin users to create
features. You can use the admin site to create a token for a user.

## Instances

The main instance is <https://osmcha.org/>. You can check the API
documentation at <https://osmcha.org/api-docs/>.

There is currently no official staging instance.

## Deployment

Check the [Deploy](DEPLOY.rst) file for instructions on how to deploy
with Heroku and Dokku.

## Get in contact

If you use, deploy or are interested in helping develop OSMCha,
subscribe to our [mailing
list](https://lists.openstreetmap.org/listinfo/osmcha-dev). You can
report errors or request new features in the [osmcha-frontend
repository](https://github.com/osmcha/osmcha-frontend).


CHANGELOG.rst
Change Log
==========

Log of changes since the 2.0 version

Check https://github.com/willemarcel/osmcha-django/releases for recent updates.

[4.0.0] - 2018-10-18

- Change feature model to be a JSONField of the Changeset Model
- Implement GDPR changes

[3.0.4] - 2018-06-29

- Fix error in changeset comments when it has special characters

[3.0.3] - 2018-06-18

- Update osmcha lib version to 0.4.8

[3.0.2] - 2018-06-05

- Update Django version to 2.0.6

[3.0.1] - 2018-06-04

- Replace oauth2 by requests_oauthlib

[3.0.0] - 2018-05-30

- Update Django to 2.0.4
- Update all requirements
- Remove unneeded dependencies

[2.7.5] - 2018-04-02

- Update social-core-auth to fix bug in authentication

[2.7.4] - 2018-04-02

- Update osmcha lib version
- Use https on OSM.org API requests
- Use SQL command instead of Django ORM in merge_reasons command

[2.7.3] - 2018-03-26

- Add 'merge_reasons' management command

[2.7.2] - 2018-03-12

- Update all dependencies, except Django 2.0
- Update osmcha lib to 0.4.6

[2.7.1] - 2018-02-07

- Fix encoding bugs in ChangesetCommentAPIView

[2.7.0] - 2018-02-01

- Disable automatic comments
- Add endpoint to allow user to post comments to changesets from inside OSMCha
- Modify user model to include message_good, message_bad and comment_feature fields

[2.6.0] - 2017-12-27

- Post a comment to OSM Changeset page when a review is added to a changeset (to use it, set the ENV variable ``DJANGO_ENABLE_CHANGESET_COMMENTS`` to True)
- Return the correct username of users in views

[2.5.0] - 2017-10-09

- Change license to BSD 2-Clause

[2.4.0] - 2017-10-02

- Handle errors when trying to create a duplicated UserWhitelist using the API
- Add 'whitelists' field to User detail endpoint

[2.3.9] - 2017-09-15

- Update osmcha lib to 0.4.3 and django-extensions

[2.3.8] - 2017-09-06

- Update django and psycopg

[2.3.7] - 2017-08-30

- Update osmcha lib to 0.4.2
- Add filter and ordering by number of Suspicion Reasons

[2.3.6] - 2017-08-29

- Update osmcha lib to 0.4.1
- Update Django Rest Framework to 3.6.4
- Documentation updates
- Hide features that doesn't have visible SuspicionReasons (to non staff users and only in the changesets endpoints)

[2.3.5] - 2017-08-09

- Disable SessionAuthentication in DRF endpoints

[2.3.4] - 2017-08-08

- General update of dependencies

[2.3.3] - 2017-08-05

- Update django version to 1.11.4
- Update documentation

[2.3.2] - 2017-07-28

- Update django version to 1.11.3
- Add RSS feed to list the changesets that matches an Area of Interest
- Add changeset_checked and changeset_ids filter to FeatureFilter
- Improve views and filter documentation
- Change date and check_date filters to accept datetime values

[2.3.1] - 2017-07-28

- Disable 'cachealot' in aws-production settings
- Update osmcha lib version to 0.4.0
- Add CHANGELOG.rst

[2.3.0] - 2017-07-25

- Add 'uid' field to BlacklistedUser model
- use 'uid' instead of 'id' in BlacklistUser views
- Allow put and patch method on BlacklistedUserDetailAPIView
- Add filter by 'uid' in list changeset and feature endpoints

[2.2.1] - 2017-07-18

- Improve 'date' and 'check_date' filters to avoid XSS
- Update documentation of AOIListCreateAPIView


[2.2.0] - 2017-07-15

- Add date field to AreaOfInterest and ordering to its list endpoint


[2.1.1] - 2017-07-14

- Add number of changesets in the stats views
- Update osmcha version to 0.3.9 in requirements


[2.1.0] - 2017-07-12

- Fix changeset and feature filters inside AoI
- Adjust django swagger authentication settings


[2.0] - 2017-06-21

- A general rewrite of osmcha-django to serve data in a REST API.


config/settings/common.py
# -*- coding: utf-8 -*-
"""
Django settings for osmcha-django project.

For more information on this file, see
https://docs.djangoproject.com/en/dev/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/dev/ref/settings/
"""
from __future__ import absolute_import, unicode_literals

import environ
import os


ROOT_DIR = environ.Path(__file__) - 3  # (/a/b/myfile.py - 3 = /)
APPS_DIR = ROOT_DIR.path('osmchadjango')

env = environ.Env()

# .env file, should load only in development environment
READ_DOT_ENV_FILE = env('DJANGO_READ_DOT_ENV_FILE', default=False)

if READ_DOT_ENV_FILE:
    # Operating System Environment variables have precedence over variables defined in the .env file,
    # that is to say variables from the .env files will only be used if not defined
    # as environment variables.
    env_file = str(ROOT_DIR.path('.env'))
    print('Loading : {}'.format(env_file))
    env.read_env(env_file)
    print('The .env file has been loaded. See common.py for more information')


# APP CONFIGURATION
# ------------------------------------------------------------------------------
DJANGO_APPS = (
    # Default Django apps:
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django.contrib.gis',

    # Useful template tags:
    # 'django.contrib.humanize',

    # Admin
    'django.contrib.admin',
)
THIRD_PARTY_APPS = (
    'rest_framework',
    'rest_framework_gis',
    'rest_framework.authtoken',
    'social_django',
    'corsheaders',
    'django_filters',
    'drf_yasg',
)

# Apps specific for this project go here.
LOCAL_APPS = (
    'osmchadjango.users',  # custom users app
    'osmchadjango.changeset',
    'osmchadjango.feature',
    'osmchadjango.supervise',
    'osmchadjango.frontend',
)

# See: https://docs.djangoproject.com/en/dev/ref/settings/#installed-apps
INSTALLED_APPS = DJANGO_APPS + THIRD_PARTY_APPS + LOCAL_APPS

# MIDDLEWARE CONFIGURATION
# ------------------------------------------------------------------------------
MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.gzip.GZipMiddleware',
    'corsheaders.middleware.CorsMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.middleware.http.ConditionalGetMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware'
]

# MIGRATIONS CONFIGURATION
# ------------------------------------------------------------------------------
MIGRATION_MODULES = {
    'sites': 'osmchadjango.contrib.sites.migrations'
}

# DEBUG
# ------------------------------------------------------------------------------
# See: https://docs.djangoproject.com/en/dev/ref/settings/#debug
DEBUG = env.bool("DJANGO_DEBUG", False)

# FIXTURE CONFIGURATION
# ------------------------------------------------------------------------------
# See: https://docs.djangoproject.com/en/dev/ref/settings/#std:setting-FIXTURE_DIRS
FIXTURE_DIRS = (
    str(APPS_DIR.path('fixtures')),
)

# EMAIL CONFIGURATION
# ------------------------------------------------------------------------------
EMAIL_BACKEND = env('DJANGO_EMAIL_BACKEND', default='django.core.mail.backends.smtp.EmailBackend')

# MANAGER CONFIGURATION
# ------------------------------------------------------------------------------
# See: https://docs.djangoproject.com/en/dev/ref/settings/#admins
ADMINS = (
    ('name', 'email@email.com'),
)

# See: https://docs.djangoproject.com/en/dev/ref/settings/#managers
MANAGERS = ADMINS

# DATABASE CONFIGURATION
# ------------------------------------------------------------------------------
# See: https://docs.djangoproject.com/en/dev/ref/settings/#databases
# DATABASES = {
#     'default': env.db('DATABASE_URL', default='postgres:///osmcha'),
# }
DATABASES = {
    'default': {
         'ENGINE': 'django.contrib.gis.db.backends.postgis',
         'NAME': 'osmcha',
         'USER': env('POSTGRES_USER'),
         'PASSWORD': env('POSTGRES_PASSWORD'),
         'HOST': env('PGHOST', default='localhost')
     }
}
DATABASES['default']['ATOMIC_REQUESTS'] = True


# GENERAL CONFIGURATION
# ------------------------------------------------------------------------------
# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# In a Windows environment this must be set to your system time zone.
TIME_ZONE = 'UTC'

# See: https://docs.djangoproject.com/en/dev/ref/settings/#language-code
LANGUAGE_CODE = 'en-us'

# See: https://docs.djangoproject.com/en/dev/ref/settings/#site-id
SITE_ID = 1

# See: https://docs.djangoproject.com/en/dev/ref/settings/#use-i18n
USE_I18N = True

# See: https://docs.djangoproject.com/en/dev/ref/settings/#use-l10n
USE_L10N = True

# See: https://docs.djangoproject.com/en/dev/ref/settings/#use-tz
USE_TZ = True

# TEMPLATE CONFIGURATION
# ------------------------------------------------------------------------------
# See: https://docs.djangoproject.com/en/dev/ref/settings/#templates
TEMPLATES = [
    {
        # See: https://docs.djangoproject.com/en/dev/ref/settings/#std:setting-TEMPLATES-BACKEND
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        # See: https://docs.djangoproject.com/en/dev/ref/settings/#template-dirs
        'DIRS': [
            str(APPS_DIR.path('templates')),
        ],
        'OPTIONS': {
            # See: https://docs.djangoproject.com/en/dev/ref/settings/#template-debug
            'debug': DEBUG,
            # See: https://docs.djangoproject.com/en/dev/ref/settings/#template-loaders
            # https://docs.djangoproject.com/en/dev/ref/templates/api/#loader-types
            'loaders': [
                'django.template.loaders.filesystem.Loader',
                'django.template.loaders.app_directories.Loader',
            ],
            # See: https://docs.djangoproject.com/en/dev/ref/settings/#template-context-processors
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.template.context_processors.i18n',
                'django.template.context_processors.media',
                'django.template.context_processors.static',
                'django.template.context_processors.tz',
                'django.contrib.messages.context_processors.messages',
                'social_django.context_processors.backends',
                'social_django.context_processors.login_redirect',
                # Your stuff: custom template context processors go here
            ],
        },
    },
]

# See: http://django-crispy-forms.readthedocs.org/en/latest/install.html#template-packs
CRISPY_TEMPLATE_PACK = 'bootstrap3'

# STATIC FILE CONFIGURATION
# ------------------------------------------------------------------------------
# See: https://docs.djangoproject.com/en/dev/ref/settings/#static-root
STATIC_ROOT = str(ROOT_DIR('staticfiles'))

# See: https://docs.djangoproject.com/en/dev/ref/settings/#static-url
STATIC_URL = '/static/'

# See: https://docs.djangoproject.com/en/dev/ref/contrib/staticfiles/#std:setting-STATICFILES_DIRS
STATICFILES_DIRS = (
    str(APPS_DIR.path('static')),
)

# See: https://docs.djangoproject.com/en/dev/ref/contrib/staticfiles/#staticfiles-finders
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
)

# MEDIA CONFIGURATION
# ------------------------------------------------------------------------------
# See: https://docs.djangoproject.com/en/dev/ref/settings/#media-root
MEDIA_ROOT = str(APPS_DIR('media'))

# See: https://docs.djangoproject.com/en/dev/ref/settings/#media-url
MEDIA_URL = '/media/'

# URL Configuration
# ------------------------------------------------------------------------------
ROOT_URLCONF = 'config.urls'

# See: https://docs.djangoproject.com/en/dev/ref/settings/#wsgi-application
WSGI_APPLICATION = 'config.wsgi.application'

# CELERY CONFIGURATION
BROKER_URL = env('CELERY_BROKER_URL', default='redis://localhost:6379/0')

# Some really nice defaults
# ACCOUNT_AUTHENTICATION_METHOD = 'username'
# ACCOUNT_EMAIL_REQUIRED = True
# ACCOUNT_EMAIL_VERIFICATION = 'mandatory'

# Custom user app defaults
# Select the correct user model
AUTH_USER_MODEL = 'users.User'

# SLUGLIFIER
AUTOSLUG_SLUGIFY_FUNCTION = 'slugify.slugify'

# SOCIAL AUTH CONFIGURATION
SOCIAL_AUTH_DEFAULT_USERNAME = lambda u: slugify(u)
SOCIAL_AUTH_ASSOCIATE_BY_EMAIL = True

SOCIAL_AUTH_OPENSTREETMAP_KEY = env('OAUTH_OSM_KEY', default='')
SOCIAL_AUTH_OPENSTREETMAP_SECRET = env('OAUTH_OSM_SECRET', default='')

# AUTHENTICATION CONFIGURATION
# ------------------------------------------------------------------------------
AUTHENTICATION_BACKENDS = (
    'social_core.backends.openstreetmap.OpenStreetMapOAuth',
    'django.contrib.auth.backends.ModelBackend',
)

SOCIAL_AUTH_PIPELINE = (
    'social_core.pipeline.social_auth.social_details',
    'social_core.pipeline.social_auth.social_uid',
    'social_core.pipeline.social_auth.auth_allowed',
    'social_core.pipeline.social_auth.social_user',
    'social_core.pipeline.user.get_username',
    'social_core.pipeline.user.create_user',
    'osmchadjango.users.utils.save_real_username',
    'social_core.pipeline.social_auth.associate_user',
    'social_core.pipeline.social_auth.load_extra_data',
    'social_core.pipeline.user.user_details'
    # 'social_core.pipeline.social_auth.associate_by_email',
)

# LOGGING CONFIGURATION
# ------------------------------------------------------------------------------
# See: https://docs.djangoproject.com/en/dev/ref/settings/#logging
# A sample logging configuration. The only tangible logging
# performed by this configuration is to send an email to
# the site admins on every HTTP 500 error when DEBUG=False.
# See http://docs.djangoproject.com/en/dev/topics/logging for
# more details on how to customize your logging configuration.

LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'handlers': {
        'console': {
            'class': 'logging.StreamHandler',
        },
    },
    'loggers': {
        'django': {
            'handlers': ['console'],
            'level': os.getenv('DJANGO_LOG_LEVEL', 'INFO'),
        },
        'osmchadjango.users': {
            'handlers': ['console', ],
            'level': "DEBUG",
        },
    },
}

# PROJECT CONFIGURATION

# If you want to filter the import of changesets to a defined area of the world,
# define CHANGESETS_FILTER as a path to a GeoJSON file.
CHANGESETS_FILTER = env('DJANGO_CHANGESETS_FILTER', default=None)

# Define your prefered visualization tool link
# Some options are 'https://nrenner.github.io/achavi/?changeset=',
# https://overpass-api.de/achavi/?changeset=
OSM_VIZ_TOOL_LINK = env('VIZ_TOOL_LINK', default='https://osmlab.github.io/changeset-map/#')

# Enable/disable the functionality to post comments to changesets when they
# are reviewed
ENABLE_POST_CHANGESET_COMMENTS = env('DJANGO_ENABLE_CHANGESET_COMMENTS', default=False)

# Authorization token to access the OSM-COMMENTS-API
OSM_COMMENTS_API_KEY = env('DJANGO_OSM_COMMENTS_API_KEY', default='')

# Your common stuff: Below this line define 3rd party library settings
REST_FRAMEWORK = {
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework.authentication.BasicAuthentication',
        'rest_framework.authentication.TokenAuthentication',
        'rest_framework.authentication.SessionAuthentication',
        ),
    'DEFAULT_SCHEMA_CLASS': 'rest_framework.schemas.coreapi.AutoSchema',
    'DEFAULT_PAGINATION_CLASS': 'rest_framework.pagination.PageNumberPagination',
    'PAGE_SIZE': 50,
    'DEFAULT_FILTER_BACKENDS': (
        'django_filters.rest_framework.DjangoFilterBackend',
        ),
    'TEST_REQUEST_DEFAULT_FORMAT': 'json',
    'DEFAULT_THROTTLE_RATES': {
        'non_staff_user': env('NON_STAFF_USER_THROTTLE_RATE', default='3/min')
        },
    'ORDERING_PARAM': 'order_by',
    }

# Allow cross domain requests
CORS_ORIGIN_ALLOW_ALL = True

# SWAGGER SETTINGS
SWAGGER_SETTINGS = {
    'USE_SESSION_AUTH': False,
    'SECURITY_DEFINITIONS': {
        'api_key': {
            'type': 'apiKey',
            'name': 'Authorization',
            'in': 'header'
            }
        },
    }

# CACHALOT SETTINGS
CACHALOT_TIMEOUT = 180
CACHALOT_ENABLED = True

# FRONTEND SETTINGS
# -----------------------------------------------------------------------------
# Version or any valid git branch tag of front-end code
OSMCHA_FRONTEND_VERSION = env('OSMCHA_FRONTEND_VERSION', default='oh-pages')

# Define the URL to where the user will be redirected after the authentication
# in OSM website
OAUTH_REDIRECT_URI = env(
    'OAUTH_REDIRECT_URI',
    default='http://localhost:8000/oauth-landing.html'
    )
